benchmarks/imagenet/resnet50/mocov2.py

```diff
@@ -0,0 +1,145 @@
+import copy
+from typing import List, Tuple
+
+import torch
+from pytorch_lightning import LightningModule
+from torch import Tensor
+from torch.nn import Identity
+from torch.optim import SGD
+from torchvision.models import resnet50
+
+from lightly.loss import NTXentLoss
+from lightly.models.modules import MoCoProjectionHead
+from lightly.models.utils import (
+    batch_shuffle,
+    batch_unshuffle,
+    get_weight_decay_parameters,
+    update_momentum,
+)
+from lightly.transforms import MoCoV2Transform
+from lightly.utils.benchmarking import OnlineLinearClassifier
+from lightly.utils.scheduler import CosineWarmupScheduler
+
+
+class MoCoV2(LightningModule):
+    def __init__(self, batch_size_per_device: int, num_classes: int) -> None:
+        super().__init__()
+        self.save_hyperparameters()
+        self.batch_size_per_device = batch_size_per_device
+
+        resnet = resnet50()
+        resnet.fc = Identity()  # Ignore classification head
+        self.backbone = resnet
+        self.projection_head = MoCoProjectionHead()
+        self.query_backbone = copy.deepcopy(self.backbone)
+        self.query_projection_head = MoCoProjectionHead()
+        self.criterion = NTXentLoss(
+            temperature=0.2,
+            memory_bank_size=(65536, 128),
+            gather_distributed=True,
+        )
+
+        self.online_classifier = OnlineLinearClassifier(num_classes=num_classes)
+
+    def forward(self, x: Tensor) -> Tensor:
+        return self.backbone(x)
+
+    @torch.no_grad()
+    def forward_key_encoder(self, x: Tensor) -> Tuple[Tensor, Tensor]:
+        x, shuffle = batch_shuffle(batch=x, distributed=self.trainer.num_devices > 1)
+        features = self.forward(x).flatten(start_dim=1)
+        projections = self.projection_head(features)
+        features = batch_unshuffle(
+            batch=features,
+            shuffle=shuffle,
+            distributed=self.trainer.num_devices > 1,
+        )
+        projections = batch_unshuffle(
+            batch=projections,
+            shuffle=shuffle,
+            distributed=self.trainer.num_devices > 1,
+        )
+        return features, projections
+
+    def forward_query_encoder(self, x: Tensor) -> Tensor:
+        features = self.query_backbone(x).flatten(start_dim=1)
+        projections = self.query_projection_head(features)
+        return projections
+
+    def training_step(
+        self, batch: Tuple[List[Tensor], Tensor, List[str]], batch_idx: int
+    ) -> Tensor:
+        views, targets = batch[0], batch[1]
+
+        # Encode queries.
+        query_projections = self.forward_query_encoder(views[1])
+
+        # Momentum update. This happens between query and key encoding, following the
+        # original implementation from the authors:
+        # https://github.com/facebookresearch/moco/blob/5a429c00bb6d4efdf511bf31b6f01e064bf929ab/moco/builder.py#L142
```

Comment:
super minor: but maybe adjust docstring length

Reply:
Not sure if we mean the same thing, but it is already adjusted (first line wraps at 90 chars) and the link is on a new line (we don't wrap links over multiple lines).